source/retryable-writes/tests/README.rst

```diff
@@ -117,6 +125,37 @@ Running these as integration tests will require a running mongod server. Each of
 these tests is valid against a standalone mongod, a replica set, and a sharded
 system for server version 3.6.0 or later.
 
+Network Error Tests
+===================
+
+The YAML tests exercise the following test scenarios:
+
+- Single-statement write operations
+  - Each test expecting a write result will encounter at-most one network error
+    for the write command. Retry attempts should return without error and allow
+    operation to succeed. Observation of the collection state will assert that
+    the write occurred at-most once.
+  - Each test expecting an error will encounter successive network errors for
+    the write command. Observation of the collection state will assert that the
+    write was never committed on the server.
+- Multi-statement write operations
+ - Each test expecting a write result will encounter at-most one network error
+   for some write command(s) in the batch. Retry attempts should return without
+   error and allow the batch to ultimately succeed. Observation of the
+   collection state will assert that each write occurred at-most once.
+ - Each test expecting an error will encounter successive network errors for
+   some write command in the batch. The batch will ultimately fail with an
+   error, but observation of the collection state will assert that the failing
+   write was never committed on the server. We may observe that earlier writes
+   in the batch occurred at-most once.
```

Comment:
Rather than duplicate these comments throughout all of the test files, I collected them here.

That said, I'm a bit bothered by the fact that we're not testing writes in the BulkWrite batch that we know succeeded before the unrecoverable failure.

I notice that the CRUD spec's [Error Handling](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#error-handling) section mentions that drivers MAY report `processedRequests` and `unprocessedRequests` on a BulkWriteException. Those are just defined as "requests that were sent to the server" and "not sent to the server", respectively, and make no mention of whether or not they succeeded.

In the PHP driver, our exception class has an [accessor method](http://php.net/manual/en/mongodb-driver-writeexception.getwriteresult.php) for the write result object up until the point of failure, which we could theoretically use to assert which operations did succeed before the unrecoverable failure. That write result would then agree with the observed collection state.

Would it make sense to add the incomplete `result:` field to the tests with `error: true`, and document that drivers may assert the result if possible?

Reply:
"Would it make sense to add the incomplete result: field to the tests with error: true, and document that drivers may assert the result if possible?"

I think that makes sense. PyMongo also has a reference to the bulk write result object in out `BulkWriteError` exception class.

The formatting of this section looks a bit off. I htink you need two spaces before the sub lists.

Reply:
Can you make that change in a separate PR though? It would be easier to review.

Reply:
Will add asserts for incomplete write results in [SPEC-971](https://jira.mongodb.org/browse/SPEC-971).

List formatting has been fixed.